### ISSUE TITLE
Boost static detection is broken

### ISSUE THREAD

**Issue by sarum9in** [mesonbuild/meson#3082]
`detect_lib_modules()` calls `find_libraries_with_abi_tags()` and falls back to `self.detect_lib_modules_win()` and `self.detect_lib_modules_nix()` only if the first one fails.
`find_libraries_with_abi_tag()` uses `compiler.find_library()` and does not respect `self.static` at all.

Ideally I would expect `compiler.find_library()` to accept `static` keyword and just pass it in this call.

Related issue #1709

**Comment by sarum9in** [comment]
#328 is a blocker for this.

**Comment by sarum9in** [comment]
Apparently `find_library()` supports `libtype` internally, which can be used here to force static or shared as needed. We can use it.

**Comment by nirbheek** [comment]
> Apparently find_library() supports libtype internally

Yeah, I added it so that other dependency types could use it. Great that someone is finally implementing it!

**Comment by bredelings** [comment]
The support for finding static libraries seems to depend on gcc --print-search-dirs, which only print system dirs.  Thus, if libraries search paths are supplied with -L, get_library_dirs() won't find them.

I think what we want to do is get GCC to search the library search dirs, but put our own prefix and suffix on the library.  You can apparently do it like this:
`g++ file.C -l:libbsd.a`

If we add this functionality, then we can use self.links( ) in compiler.find_library( ), which seem like the correct way to do things.

**Comment by bredelings** [comment]
The self.detect_lib_modules_win() and self.detect_lib_modules_nix() is old code that has a number of problems.  One problem is that when it is searching for library files in a specific directory it does _not_ consider all the relevant prefixes and suffixes like compiler.find_library( ) does near the node "Triply nested loop!"

**Comment by bredelings** [comment]
Another way that self.detect_lib_modules_win() and self.detect_lib_modules_nix() could be improved is that we know the list of boost libraries that we are looking for, so we shouldn't be using globs for that.  We should probably just compute the name of the library as boost_<module>-<abi-tag> and then do a loop over all prefixes and suffixes to see if it exists.

**Comment by bredelings** [comment]
However, ultimately, I am hoping that we can simplify self.detect_lib_modules_win() and self.detect_lib_modules_nix() to the point where they disappear.  There are lot of broken hacks in there that do not generally work.

**Comment by jpakkane** [comment]
It is my sincerest hope that once Boost finishes transitioning to CMake, they will start shipping pkg-config files.

**Comment by bredelings** [comment]
The discussion on this ticket seems very relevant:
https://svn.boost.org/trac10/ticket/1094

>The point is that pkg-config doesn't solve the "I don't know which configuration I want" problem. It
> solves the "I don't know what compiler/linker flags I need" problem while assuming that the .pc file
> name you provided corresponds to an acceptable configuration.

It seems that the issue is selecting between different installed variants of the library that have unguessable names.  We must make them stop doing this! :-)

However, it seems that this all comes from the fact that windows uses different crt object for the debug and non-debug versions of a library.  So, on windows, some debug and non-debug versions of a library often have different names, and this is not limitted to boost.

Maybe we could get the pkg-config people to address the issue of having multiple library variants?

**Comment by bredelings** [comment]
I am not sure that `-l:libname.a` works in a cross-platform fashion...

**Comment by sarum9in** [comment]
I think if we just use full pathname to libname.a it will just work. Currently work here is blocked by #3104 anyway. Test cases for boost are simply wrong at the moment and we need to fix them as well. They "verify" that static linking is working but they just link dynamically. My idea is to use boost::log, because of its name mangling it will definitely fail to build if linking is not matching header definitions.

**Comment by bredelings** [comment]
My concern is that this won't work in situations where we can't find the full pathname, like with macports.  We could just say that forcing static linking of boost is not supported in those cases I suppose.

**Comment by nirbheek** [comment]
> My concern is that this won't work in situations where we can't find the full pathname, like with macports

You can pass extra library paths to `find_library()`, so if we know where the location is in such cases, we can use that.

**Comment by bredelings** [comment]
This is true, but its not clear that we do know where the location is.  If we specify a library search path with -L, then the compiler knows to search this, but meson does not.

**Comment by nirbheek** [comment]
> If we specify a library search path with -L, then the compiler knows to search this

Where does this path come from?

**Comment by bredelings** [comment]
From LDFLAGS and/or from `link_args`.

**Comment by nirbheek** [comment]
It definitely makes sense to have `find_library()` look in `LDFLAGS`. If by `link_args` you mean on a target, that's not known at the time of dependency resolution, and is absolutely the wrong way to try to find a dependency.

**Comment by nirbheek** [comment]
If you want to use a Boost that's in a standard system location like Macports or on Windows, those paths should be encoded in the list of `BoostDependency` search paths.

**Comment by bredelings** [comment]
Sure, linker arguments need to be known at the time of dependency resolution.  In addition to LDFLAGS, this could include:
- add_project_link_arguments( )
- cpp_link_args property from a cross-file.
I don't know if there are other places.

**Comment by bredelings** [comment]
I don't think the idea of a "standard location" works with macports.  You could have libraries installed under /usr/local/ for homebrew, under /opt/local for macports and maybe under /usr/ for apple.  In this situation a user needs to specify a search order using -I and -L flags in CPPFLAGS and LDFLAGS.

In general, my two concerns are:
- I think we need to allow user-supplied search paths for include files and libraries
- Trying to find library files ourselves seems like we are trying to duplicate the functionality of the compiler inside meson, and this might be fragile.  If we use self.links( ) then this will "just work".

As an example of fragility, I thought that some windows compilers uses \LIBPATH instead of '-L'.

**Comment by nirbheek** [comment]
> Trying to find library files ourselves seems like we are trying to duplicate the functionality of the compiler inside meson, and this might be fragile. If we use self.links( ) then this will "just work".

`compiler.find_library()` does use `self.links()`, and with linkers that support it, we can tell it to search for static libraries with `-l:libfoo.a` syntax. However, we always need to search manually because not all linkers support that syntax.

**Comment by bredelings** [comment]
I agree that this seems like the best approach for now.  If we could find some way to get the functionality of libfoo.a in a cross-platform way, this would be even better, but I don't see how to do that yet.
